src/sdk/namespace_fs.js

```diff
@@ -492,23 +494,94 @@ class NamespaceFS {
+        return { etag: this._get_etag(stat) };
+    }
 
-            await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
-            await this._upload_stream(params.source_stream, upload_path, fs_account_config);
-            // TODO use file xattr to store md5_b64 xattr, etc.
-            const stat = await nb_native().fs.stat(fs_account_config, upload_path);
-            await nb_native().fs.rename(fs_account_config, upload_path, file_path);
-            return { etag: this._get_etag(stat) };
-        } catch (err) {
-            throw this._translate_object_error_codes(err);
+    // Comparing both device and inode number (st_dev and st_ino returned by stat) 
+    // will tell you whether two different file names refer to the same thing.
+    // If so, we will return the etag of the file_path
+    async _is_same_inode(fs_account_config, source_file_path, file_path) {
+        try {
+            const file_path_stat = await nb_native().fs.stat(fs_account_config, file_path);
+            const file_path_inode = file_path_stat.ino.toString();
+            const file_path_device = file_path_stat.dev.toString();
+            const source_file_stat = await nb_native().fs.stat(fs_account_config, source_file_path);
+            const source_file_inode = source_file_stat.ino.toString();
+            const source_file_device = source_file_stat.dev.toString();
+            if (file_path_inode === source_file_inode && file_path_device === source_file_device) {
+                return { etag: this._get_etag(file_path_stat) };
+            }
+        } catch (e) {
+            return false;
+        }
+        return false;
+    }
+
+    async _copy_stream(source_file_path, upload_path, fs_account_config) {
+        let target_file;
+        let source_file;
+        try {
+            //Opening the source and target files
+            source_file = await nb_native().fs.open(fs_account_config, source_file_path);
+            target_file = await nb_native().fs.open(fs_account_config, upload_path, 'w');
+            //Reading the source_file and writing into the target_file
+            const { buffer } = await buffers_pool.get_buffer(config.NSFS_BUF_SIZE);
+            let read_pos = 0;
+            for (;;) {
+                const bytesRead = await source_file.read(fs_account_config, buffer, 0, config.NSFS_BUF_SIZE, read_pos);
+                if (!bytesRead) break;
+                read_pos += bytesRead;
+                const data = buffer.slice(0, bytesRead);
+                await target_file.write(fs_account_config, data);
+            }
+            //Closing the source_file and target files
+            await source_file.close(fs_account_config);
+            source_file = null;
+            await target_file.close(fs_account_config);
+            target_file = null;
+        } catch (e) {
+            console.error('Failed to copy object', e);
```

Comment:
Why you don't throw the error here? I would expect to have a throw when we did not succeed to copy.